### PR TITLE
feat: 'End Session' button (GH-5012)

### PR DESCRIPTION
**Pull request by qjimmy**
Closes GH-5012

### CODE REVIEW

File: projects/storefrontlib/src/cms-components/asm/asm-main-ui/asm-main-ui.component.spec.ts [hunk 1]
```diff
@@ -287,4 +297,34 @@ describe('AsmMainUiComponent', () => {
...
     });
   });
+
+  it("should end session on 'End Session' button click", () => {
+    //agent sign in
+    const agent_login = {
+      userId: 'asagent',
+      password: 'password',
+    };
+    component.loginCustomerSupportAgent(agent_login);
```
Comment (plabadie) [inline projects/storefrontlib/src/cms-components/asm/asm-main-ui/asm-main-ui.component.spec.ts]: Is this part required?

Comment (qjimmy) [reply projects/storefrontlib/src/cms-components/asm/asm-main-ui/asm-main-ui.component.spec.ts]: I don't think so, I'll try to change it.

File: projects/storefrontlib/src/cms-components/asm/asm-main-ui/asm-main-ui.component.spec.ts [hunk 2]
```diff
@@ -287,4 +297,49 @@ describe('AsmMainUiComponent', () => {
...
+    fixture.detectChanges();
+
+    //Click button
+    const endSessionButton = fixture.debugElement.query(
+      By.css('.fd-button--negative')
+    );
+    spyOn(routingService, 'go');
+    endSessionButton.nativeElement.click();
+
+    expect(routingService.go).toHaveBeenCalled();
```
Comment (plabadie) [inline projects/storefrontlib/src/cms-components/asm/asm-main-ui/asm-main-ui.component.spec.ts]: This does not look like it asserts that the storefront is routed back to the home page.

Is there a way we can assert the parameter the router was called with?

File: projects/storefrontlib/src/cms-components/asm/asm-main-ui/asm-main-ui.component.spec.ts [hunk 3]
```diff
@@ -87,6 +88,13 @@ class MockRoutingService {
   go() {}
 }
 
+class MockIconLoaderService {
+  getSvgPath() {}
+  useSvg() {}
+  getStyleClasses() {}
+  addLinkResource() {}
```
Comment (plabadie) [inline projects/storefrontlib/src/cms-components/asm/asm-main-ui/asm-main-ui.component.spec.ts]: Do all these functions need to be mocked?

Comment (qjimmy) [reply projects/storefrontlib/src/cms-components/asm/asm-main-ui/asm-main-ui.component.spec.ts]: getSvgPath() actually does not need to be mocked. Thanks for pointing it out!

File: projects/storefrontapp-e2e-cypress/cypress/helpers/asm.ts [hunk 4]
```diff
@@ -200,6 +229,15 @@ export function asmTests() {
...
     });
+
+    describe('End Session Button', () => {
+      it('should display the asm menu for an agent when ?asm=true is passed into the url', () => {
+        checkout.visitHomePage('asm=true');
+        cy.get('cx-asm').should('exist');
+        cy.get('cx-asm-main-ui').should('exist');
+        cy.get('cx-csagent-login-form').should('exist');
+      });
+    });
```
Comment (plabadie) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/asm.ts]: Looks like a duplicate of the test we have on line 28 
`agent should see the asm UI when ?asm=true is passed to the url'`

File: projects/storefrontapp-e2e-cypress/cypress/helpers/asm.ts [hunk 5]
```diff
@@ -149,6 +149,35 @@ export function asmTests() {
...
+
+        cy.get('cx-customer-selection').should('not.exist');
+        cy.get('div.cx-customer-emulation').should('exist');
+
+        // should end a user's session when clicking on the end session button
+        cy.get('div.cx-customer-emulation button').click();
+        cy.get('div.cx-customer-emulation').should('not.exist');
+        cy.get('cx-customer-selection').should('exist');
+      });
+
```
Comment (plabadie) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/asm.ts]: The test for the end session button should rather be after we end the session ( test 'agent should stop customer emulation.' ) but before we sign out the agent.
This way the test will be simpler.  We only need to start a customer emulation and end it right away.

File: projects/storefrontapp-e2e-cypress/cypress/helpers/asm.ts [hunk 6]
```diff
@@ -222,5 +207,51 @@ export function listenForUserDetailsRequest(): string {
...
+  });
+
+  cy.wait(aliasRequest)
+    .its('status')
+    .should('eq', 200);
+  cy.get('cx-csagent-login-form').should('not.exist');
+  cy.get('cx-customer-selection').should('exist');
+};
+
+export const customerAuthentication = fullname => {
```
Comment (plabadie) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/asm.ts]: Lets discuss the full name parameter and the function syntax.  Also, this was probably added by me previously, but there is something fishy about using the asm. prefix to call functions.

**Pull request merged**